gen-docs.ts

```diff
@@ -0,0 +1,130 @@
+    const ARG_TYPE_PATTERN = /(\w+):\s+as_(\w+)/g;
+
+    const widgets: Record<string, Widget> = {};
+    const lines = code.split("\n") 
+    
+    let currentWidget = "";
+    
+    for (let lineIndex = 0; lineIndex < lines.length; lineIndex++) {
+        
+        const line = lines[lineIndex]
+        const newWidgetMatch = NEW_WIDGET_PATTERN.exec(line);
+
+        if (newWidgetMatch && newWidgetMatch.length >= 3) {
+            const name = newWidgetMatch[2];
+            /** @type string[] */
+            const exts = newWidgetMatch[3]
+                ? newWidgetMatch[3].split(/, */)
+                : [];
+            currentWidget = name;
+            widgets[currentWidget] = {
+                name,
+                exts,
+                desc: "",
+                props: [],
+                isVisible: newWidgetMatch[1] !== "!",
+            };
+            continue;
+        }
+
+        const descMatch = line.match(DESC_PATTERN);
+        if (descMatch && descMatch.length == 2) {
+            widgets[currentWidget].desc = descMatch[1];
+            continue;
+        }
+
+        const propMatch = line.match(PROP_PATTERN);
+        if (propMatch && propMatch.length == 3) {
+            let no = lineIndex + 1
+           
+            while (/\s+\/\//.test(lines[no])) {
+                no += 1
+            } // continue till you find the line with actual code
+            
+            const matches = [...lines[no].matchAll(ARG_TYPE_PATTERN)].map(z => { z.shift(); return z }).flat() // try to use the iterator directly
+            let type = null
+            
+            const possibleMatch = matches.findIndex(x => x == propMatch[1].replaceAll("-", "_"))
+            if (possibleMatch == -1) {
+                console.log(`Failed to find a match for "${propMatch[1].replace("-", "_")}" ~ ${JSON.stringify(matches, null, 2)} ~ ${lines[no]}`)
+            }
+            type = matches[possibleMatch + 1]
+
+            widgets[currentWidget].props.push({
+                name: propMatch[1],
+                desc: propMatch[2],
+                type: type || "no-type-found"
+            });
+        }
+    }
+    return widgets;
+}
+
+function printDocs(vars: Record<string, string>, docs: Record<string, Widget>) {
+    const output = Object.values(docs)
+        .filter((x) => x.isVisible)
+        .map((x) => {
+            x.props = [
+                ...x.props,
+                ...x.exts.map((w) => docs[w]).flatMap((w) => w.props),
+            ];
+            return x;
+        })
+        .map((x) => printWidget(x))
+        .map((x) => x.replace(/\$\w+/g, (x) => vars[x.replace("$", "")]))
+        .join("\n\n");
+    console.log(output);
+}
+
+function printWidget(widget: Widget) {
+    return `
+## \`${widget.name}\` ${widget.desc ? `\n${widget.desc}` : ""}
+
+**Properties**
+${widget.props.map((prop) => `- [${prop.type}] **\`${prop.name}\`**: ${prop.desc}`).join("\n")}
```

Comment:
I'd prefer a format like
```
` - **\`${prop.name}\`**: _\`${prop.type}\`_ ${prop.desc}`
```

Something like this:
![image](https://user-images.githubusercontent.com/5300871/116925823-fbb5e200-ac59-11eb-9ac4-2c54f2cf23be.png)